src/postgraphile/withPostGraphileContext.ts

```diff
@@ -549,18 +550,34 @@ export function debugPgClient(pgClient: PoolClient, allowExplain = false): PoolC
               }
               // Explain it
               const explain = `${explainCommand} ${query}`;
+              explainPromise = pgClient[$$pgClientOrigQuery]
+                // Create a savepoint before running the EXPLAIN, so we can roll back to it to avoid running mutations
+                // twice when ANALYZE is enabled.
+                .call(this, 'savepoint postgraphile_explain')
+                .then(
+                  // Savepoint created - we are in a transaction, which means this is a mutation. We need to roll back
+                  // to the savepoint after running the EXPLAIN.
+                  () => pgClient[$$pgClientOrigQuery].call(this, explain, values)
+                    .then((data: any) => pgClient[$$pgClientOrigQuery]
+                      .call(this, 'rollback to savepoint postgraphile_explain')
+                      .then(() => data)),
+                  // Failed to create savepoint - we are not in a transaction, which means this is a query. No need to
+                  // rollback the EXPLAIN.
+                  () => pgClient[$$pgClientOrigQuery].call(this, explain, values)
+                )
+                .then((data: any) => data.rows)
+                // swallow errors during explain
+                .catch(() => null);
```

Comment:
This isn't safe because our queries can occur in parallel, and thus the rollback might be issued out of order (or there may even be concurrent rollbacks for the same savepoint!)

Instead you must ensure that they are all added in the same tick so that they get issued in order without a chance for other (parallel) queries to be inserted in between:

```js
// IMPORTANT: the following 3 lines MUST BE CALLED SYNCHRONOUSLY.
pgClient[$$pgClientOrigQuery].query('savepoint postgraphile_explain');
const explainResultPromise = pgClient[$$pgClientOrigQuery].query(explain, values);
pgClient[$$pgClientOrigQuery].query('rollback savepoint postgraphile_explain');

// Now us `explainResultPromise` however you want.
```